feat(assets): add aggregated balance selectors for account group and wallets

amitabh94 opened the pull request:
## **Description**

Integrate aggregated balance and balance change for the selected account group into Wallet Overview (state2), with privacy-mode support and consistent formatting.

- Add `AccountGroupBalance` to render the selected account group's total fiat balance with `SensitiveText`, locale-aware currency formatting, and privacy toggle.
- Add `AccountGroupBalanceChange` to render amount and percent change for a period (`1d`) for the selected account group, color-coded by sign and hidden in privacy mode.
- Wire components into `CoinOverview` under the state2 unified view flag, replacing legacy inline rendering; preserve legacy/EVM and non‑EVM paths when the flag is off.
- Introduce Redux selectors in `ui/selectors/assets.ts` that compose minimal controller state and delegate aggregation to `@metamask/assets-controllers`:
  - `selectBalanceForAllWallets`
  - `selectBalanceChangeForAllWallets(period)`
  - `selectBalanceChangeByAccountGroup(groupId, period)`
  - `selectBalanceChangeBySelectedAccountGroup(period)`
  - `selectBalanceBySelectedAccountGroup`
- Update account list to support formatted balances per group (non-breaking).


> **P.S. This PR doesn't include the native and staked balances . This is being tackled in another PR [here](https://github.com/MetaMask/metamask-extension/pull/35540) which bumps the asset controllers package that includes native and staked balances logic.**

## **Changelog**

CHANGELOG entry: null

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/ASSETS-1126

## **Manual testing steps**

```
State2 Unified View (selected account group):
1. Ensure feature flag is enabled for Multichain Accounts state2.
2. Open Wallet Overview on Home.
3. Verify the main balance shows the selected group’s fiat total with a privacy toggle.
4. Toggle Privacy Mode in Settings and confirm the balance value and suffix hide.
5. Confirm a `1d` change row shows amount (with +/−) and percent in parentheses, color-coded (green for gains, red for losses); values are hidden in privacy mode.

Fallback paths (flag off):
6. With flag disabled, confirm legacy EVM aggregate or non‑EVM aggregate still renders unchanged.

Formatting/locale:
7. Change display currency and locale; confirm currency code and number formatting update accordingly.

Edge cases:
8. With no selected account group, balance renders a spinner then resolves once selected.
9. With zero balances, amount shows “+$0.00 (0.00%)” style, colored neutral.

Account list (balances):
10. Open the Multichain Account List (menu or sidebar).
11. Verify each account group displays its fiat total balance; values are hidden in privacy mode.
12. Switch to another account group and confirm the list selection updates and Wallet Overview main balance and 1d change reflect the new group.
13. Change display currency and locale; confirm list balances reformat accordingly.
14. With zero balances, balances show "$0.00"; with the flag off, legacy list behavior remains unchanged.
```

## **Screenshots/Recordings**

### **Before**

**With BIP44 feature flag OFF**

https://github.com/user-attachments/assets/8fca2682-84b2-4507-b8e1-d18fdb50f843



### **After**

**With BIP44 feature flag ON**

https://github.com/user-attachments/assets/3efc5f73-7bc8-4953-931d-53e6647d0933



## **Pre-merge author checklist**
- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**
- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

amitabh94 committed:
feat(assets): gate home balance by State2 flag and use aggregated selector when enabled

amitabh94 committed:
chore: update @metamask/assets-controllers to version 73.1.0 and @metamask/keyring-api to version 20.0.0

amitabh94 committed:
style(assets): lint/order fixes for coin-overview gating edits

amitabh94 commented:
I have read the CLA Document and I hereby sign the CLA

amitabh94 committed:
feat(assets): use assets-controllers balance selectors for State2 aggregated home balance; align PR template

amitabh94 committed:
fix(assets): use selectBalanceForSelectedAccountGroup for State2 home balance

amitabh94 committed:
chore(assets): add debug logs for legacy vs State2 aggregated home balance

amitabh94 committed:
chore: remove debug logs per review

amitabh94 committed:
fix(assets): guard State2 selector when AccountTreeController not initialized

amitabh94 committed:
fix(assets): guard all-wallets and group resolution until AccountTreeController is ready

amitabh94 committed:
fix(assets): prefer selected-group selector; add mobile-parity group resolution; safe portfolio fallback with currentCurrency

amitabh94 committed:
fix(assets): read selectedAccountGroup from UI state (metamask.accountTree) and guard tree access; use assets balance selectors for selected group

amitabh94 committed:
fix(assets): integrate new balance selectors for multichain accounts; enhance portfolio change calculations and improve state management

amitabh94 committed:
chore(assets): update assets-controllers dependency to npm package version 73.1.0-preview-343f3d69

amitabh94 committed:
feat(assets): implement AccountGroupBalanceChange component and integrate it into CoinOverview for displaying portfolio changes

amitabh94 committed:
chore: remove deprecated @metamask/base-controller dependency from yarn.lock

amitabh94 committed:
chore: update @metamask/assets-controllers to version 73.1.0-preview-a89812f5 and bump @metamask/controller-utils to version 11.12.0

amitabh94 committed:
refactor: rename portfolio change selectors to balance change selectors for consistency

Updated selector names and their usages across the codebase to reflect the change from portfolio to balance terminology. This includes adjustments in the `account-group-balance-change` component, `aggregated-balance-state2`, and related tests to ensure clarity and maintainability.

amitabh94 committed:
refactor(assets): streamline AggregatedBalanceState2 component and update balance selectors

Refactored the AggregatedBalanceState2 component to simplify its logic by removing unnecessary state calculations and relying solely on the selected group balance. Updated the selector names from `selectAggregatedBalanceByAccountGroup` to `selectBalanceByAccountGroup` for consistency across the codebase. This change enhances clarity and maintainability in the balance management logic.

amitabh94 committed:
Merge branch 'main' into ASSETS-1126-integrate-aggregated-balance-selector

amitabh94 committed:
chore: update @metamask/assets-controllers to version 73.2.0

amitabh94 committed:
Merge branch 'main' into ASSETS-1126-integrate-aggregated-balance-selector

amitabh94 committed:
chore: update @metamask/assets-controllers to version 73.3.0 and @metamask/keyring-api to version 20.1.0

amitabh94 committed:
Merge branch 'main' into ASSETS-1126-integrate-aggregated-balance-selector

amitabh94 committed:
Merge branch 'main' into ASSETS-1126-integrate-aggregated-balance-selector

amitabh94 committed:
chore: update @metamask/keyring-api version in yarn.lock

This commit consolidates the version entries for @metamask/keyring-api, updating it to version 20.1.0 while removing the previous version 20.0.0 entry. This change helps streamline dependency management in the project.

amitabh94 committed:
chore: apply patch for @metamask/assets-controllers version 73.3.0

This commit updates the @metamask/assets-controllers dependency in package.json to use a patch version, ensuring that the specific patch file is applied. The corresponding changes in yarn.lock reflect this update, maintaining consistency in dependency management across the project.

amitabh94 committed:
Merge branch 'main' into ASSETS-1126-integrate-aggregated-balance-selector

amitabh94 committed:
refactor: remove unused getTokensMarketData import and console log in coin-overview component

amitabh94 committed:
refactor: enhance balance selectors with detailed JSDoc comments

This commit adds comprehensive JSDoc comments to various balance selectors in the assets selector file. The comments clarify the purpose and parameters of each selector, improving code readability and maintainability. This change supports better documentation practices and aids future developers in understanding the balance computation logic.

amitabh94 committed:
chore: update @metamask/assets-controllers to version 74.3.1 and remove obsolete patch

This commit updates the @metamask/assets-controllers dependency from a patched version (73.3.0) to the latest stable version (74.3.1) in both package.json and yarn.lock. The obsolete patch file for the previous version has been removed, streamlining the dependency management and ensuring the use of the latest features and fixes.

amitabh94 committed:
refactor: update balance selectors to remove percent change and add new account group selector

This commit refactors the balance selectors by removing the `selectBalancePercentChange` function, aligning with the mobile API surface. It introduces `selectBalanceChangeBySelectedAccountGroup`, which returns null when no account group is selected, enhancing the clarity and functionality of balance calculations. Additionally, JSDoc comments have been added to improve documentation and maintainability.

amitabh94 committed:
feat: add AccountGroupBalance component for displaying account group balances

This commit introduces the `AccountGroupBalance` component, which displays the total balance of the selected account group in the user's currency. It utilizes Redux selectors to fetch the necessary data and includes a sensitive toggle for privacy mode. The component is designed to enhance the user experience by providing clear visibility of account group balances while adhering to the existing design system.

amitabh94 committed:
test: add unit tests for AccountGroupBalance component

This commit introduces unit tests for the `AccountGroupBalance` component, ensuring it renders correctly under various conditions. The tests cover scenarios such as displaying a loading spinner when no balance is selected and rendering the formatted balance and currency when data is available. This addition enhances the reliability of the component and supports ongoing development efforts.

amitabh94 committed:
refactor: update AccountGroupBalanceChange component to use new balance selector and formatting utility

This commit refactors the `AccountGroupBalanceChange` component by replacing the old balance selectors with the new `selectBalanceChangeBySelectedAccountGroup` selector. It also removes the custom formatting logic in favor of the `formatWithThreshold` utility for better consistency and readability. The changes enhance the component's performance and maintainability while ensuring accurate display of account group balance changes.

amitabh94 committed:
refactor: streamline CoinOverview component by integrating AccountGroupBalance and optimizing render logic

This commit refactors the `CoinOverview` component to utilize the new `AccountGroupBalance` component for displaying account group balances. The render logic has been optimized to reduce complexity, particularly in the `renderAggregatedView` function, by simplifying conditional rendering. Additionally, the early exit for the state2 unified view has been clarified, enhancing code readability and maintainability while ensuring accurate balance display across different account states.

amitabh94 committed:
test: enhance unit tests for AccountGroupBalanceChange component to cover privacy mode scenarios

This commit updates the unit tests for the `AccountGroupBalanceChange` component to include scenarios for the new privacy mode feature. It ensures that the component correctly renders masked values when privacy mode is enabled and displays the appropriate amount and percentage when privacy mode is disabled. These enhancements improve test coverage and reliability, supporting ongoing development efforts.

amitabh94 committed:
test: update unit tests for AccountGroupBalance component to use typed mock balance

amitabh94 committed:
Merge branch 'main' into ASSETS-1126-integrate-aggregated-balance-selector

amitabh94 committed:
Revert "chore: update @metamask/assets-controllers to version 74.3.1 and remove obsolete patch"

This reverts commit 63f9e611272e25b99d1d31f2f2b62adec37b1faf.

amitabh94 committed:
Merge branch 'main' into ASSETS-1126-integrate-aggregated-balance-selector

amitabh94 committed:
refactor: update file to keep only necessary changes

amitabh94 committed:
chore: move multichain account list UI changes to split PR

amitabh94 committed:
Merge branch 'main' into ASSETS-1126-integrate-aggregated-balance-selector

amitabh94 committed:
feat: integrate balance selectors with account list

Prithpal-Sooriya commented on ui/components/app/assets/account-group-balance-change/account-group-balance-change.tsx:
```diff
@@ -0,0 +1,147 @@
+    }
+  }
+
+  const percentNumber =
+    typeof portfolioPercent === 'number' ? portfolioPercent : 0;
+  const signPrefix = percentNumber >= 0 ? '+' : '-';
+  const absPercent = Math.abs(percentNumber);
+  const displayAbsPercent =
+    absPercent !== 0 && absPercent < 0.01 ? 0.01 : absPercent;
+  const localizedPercent = Intl.NumberFormat(locale, {
+    style: 'percent',
+    maximumFractionDigits: 2,
+    minimumFractionDigits: 2,
+  }).format(displayAbsPercent / 100);
+  const formattedPercentWithParens = `(${signPrefix}${localizedPercent})`;
+
+  let formattedAmount = '';
+  if (
+    portfolioChange &&
+    isValidAmount(portfolioChange.amountChangeInUserCurrency)
+  ) {
+    const amt = Number(portfolioChange.amountChangeInUserCurrency);
+    const signPrefixAmount = amt >= 0 ? '+' : '';
+    try {
+      formattedAmount = `${signPrefixAmount}${Intl.NumberFormat(locale, {
+        style: 'currency',
+        currency: fiatCurrency,
+        minimumFractionDigits: 2,
+        maximumFractionDigits: 2,
+      }).format(amt)} `;
+    } catch {
+      formattedAmount = `${signPrefixAmount}${Intl.NumberFormat(locale, {
+        style: 'decimal',
+        minimumFractionDigits: 2,
+        maximumFractionDigits: 2,
+      }).format(amt)} `;
+    }
+  }
```
Is this not something that already exists as a util?
I would prefer moving this into a utility.

Prithpal-Sooriya replied:
We can test in isolation and not tied to components/state.

david0xd reviewed (changes requested):
Some changes requested. Some stuff needs to be clarified.

Prithpal-Sooriya committed:
Merge branch 'main' of github.com:MetaMask/metamask-extension into ASSETS-1126-integrate-aggregated-balance-selector

Prithpal-Sooriya commented on ui/components/app/assets/account-group-balance-change/account-group-balance-change.tsx:
```diff
@@ -0,0 +1,147 @@
+import { selectBalanceChangeBySelectedAccountGroup } from '../../../../selectors/assets';
+import { Box, SensitiveText } from '../../../component-library';
+
+// Simple inline implementations to avoid restricted imports
```
Removing this comment, not needed

Prithpal-Sooriya commented on ui/components/app/assets/account-group-balance-change/account-group-balance-change.tsx:
```diff
@@ -0,0 +1,147 @@
+const isValidAmount = (value: unknown): value is number =>
+  typeof value === 'number' && !Number.isNaN(value) && Number.isFinite(value);
+
+// removed: replaced by formatWithThreshold-based formatting
```
Removing this comment, not needed

Prithpal-Sooriya replied:
Cleaned up in d218940b9f8970881e8aedf741bbaa7066571c43

Prithpal-Sooriya commented on ui/components/app/assets/account-group-balance-change/account-group-balance-change.tsx:
```diff
@@ -0,0 +1,147 @@
+// removed: replaced by formatWithThreshold-based formatting
+
+export type AccountGroupBalanceChangeProps = {
+  period: '1d' | '7d' | '30d';
```
hmmmmm, I'm open to both unions and enums.
As this is tied only to this component, I'm open to use an enum.

Prithpal-Sooriya committed:
refactor: clean up account-group-balance-change component

split out utils; hooks; and add comprehensive tests

Prithpal-Sooriya commented on ui/components/app/assets/account-group-balance/account-group-balance.tsx:
```diff
@@ -0,0 +1,102 @@
+    <>
+      <Box
+        className={classnames(`${classPrefix}-overview__primary-balance`, {
+          [`${classPrefix}-overview__cached-balance`]: balanceIsCached,
```
Actually I take that back... it looks like this is used in the coin-overview component.
I don't like this classname pass down, it makes classname style selectors very unreliable.

Prithpal-Sooriya committed:
refactor: clean up account-group-balance component and test

Prithpal-Sooriya commented on ui/pages/multichain-accounts/account-list/account-list.tsx:
Wow this isn't virtualised???

Prithpal-Sooriya committed:
refactor: create and use wallet and account balance hooks

provides an abstraction for getting balances for accounts and per wallet. I think we can also investigate making out selectors more fine grain to prevent additional rerenders

Prithpal-Sooriya commented on ui/hooks/multichain-accounts/useAccountBalance.ts:
```diff
@@ -0,0 +1,84 @@
+    [],
+  );
+
+  return displayBalanceCalc;
```
Agreed, I'm a little annoyed that my local VSCode setup is not picking up linting errors 😠

Prithpal-Sooriya committed:
refactor: fix linting

Prithpal-Sooriya commented on ui/pages/multichain-accounts/account-list/account-list.tsx:
```diff
@@ -1,4 +1,5 @@
-import React from 'react';
```
Update the Wallet Details Page balances

Prithpal-Sooriya replied:
Added support for this in d92dc3015d5b2412375608445da36303cd70fbe0

Prithpal-Sooriya commented on ui/components/multichain-accounts/multichain-account-list/multichain-account-list.tsx:
```diff
@@ -120,6 +126,12 @@ export const MultichainAccountList = ({
         const groupsItems = Object.entries(walletData.groups || {}).flatMap(
           ([groupId, groupData]) => {
+            // If prop is provided, attempt render balance. Otherwise do not render balance.
+            const balanceText = formattedAccountGroupBalancesByWallet
+              ? (formattedAccountGroupBalancesByWallet?.[walletId]?.[groupId] ??
+                '$ n/a')
+              : '';
```
needs confirmation - do we really want to show 'n/a' all the time as a fallback? Esp if this prop isn't passed in?

gantunesr replied:
We shouldn't use `$ n/a`. As @david0xd mentioned, this was just a placeholder during development, we should fallback to whatever value we use across the app for balances so we'll take your suggestion on this one @Prithpal-Sooriya.

Prithpal-Sooriya committed:
refactor: comment cleanup

Prithpal-Sooriya commented on ui/components/app/assets/account-group-balance/account-group-balance.tsx:
```diff
@@ -0,0 +1,102 @@
+          onClick={handleSensitiveToggle}
+          iconName={privacyMode ? IconName.EyeSlash : IconName.Eye}
+          justifyContent={JustifyContent.center}
+          ariaLabel="Sensitive toggle"
```
Cleaned up in d218940b9f8970881e8aedf741bbaa7066571c43

Prithpal-Sooriya commented on ui/components/app/assets/account-group-balance-change/account-group-balance-change.tsx:
```diff
@@ -0,0 +1,147 @@
+  const locale = useSelector(getIntlLocale) as string;
+  const { privacyMode } = useSelector(getPreferences);
+
+  // Memoized selector for the specified period
```
Lets pencil this, I do agree that cache is wiped here and we need to see impact of this when measuring.

In mobile performance research for parameterised reselect selectors we used weakmap caching.
https://github.com/MetaMask/metamask-mobile/blob/adb0c51ae14f950b8d9ae59b3355ae2f41332812/app/selectors/tokenBalancesController.ts#L36-L62

Prithpal-Sooriya commented on ui/components/multichain-accounts/multichain-account-list/multichain-account-list.tsx:
```diff
@@ -120,6 +126,12 @@ export const MultichainAccountList = ({
+              ? (formattedAccountGroupBalancesByWallet?.[walletId]?.[groupId] ??
+                '$ n/a')
+              : '';
+
```
Yeah did you not see my comment above on this?

If you don't provide a prop, does it make sense to still show `'$ n/a'`?

Prithpal-Sooriya committed:
refactor: fix type error

Prithpal-Sooriya committed:
Merge branch 'main' of github.com:MetaMask/metamask-extension into ASSETS-1126-integrate-aggregated-balance-selector

Prithpal-Sooriya commented on ui/components/multichain-accounts/multichain-account-list/multichain-account-list.tsx:
```diff
@@ -41,12 +41,17 @@ export type MultichainAccountListProps = {
+  formattedAccountGroupBalancesByWallet?: Record<
+    string,
+    Record<string, string>
+  >;
```
Nice. The plan will be to get these merged and next week can be dedicated to performance testing and improvements 😄

bergarces committed:
Merge branch 'main' into ASSETS-1126-integrate-aggregated-balance-selector

bergarces committed:
Merge branch 'main' into ASSETS-1126-integrate-aggregated-balance-selector

bergarces committed:
fix test

salimtb committed:
Merge branch 'main' into ASSETS-1126-integrate-aggregated-balance-selector

salimtb committed:
fix: fix after merge

salimtb committed:
Merge branch 'main' into ASSETS-1126-integrate-aggregated-balance-selector

bergarces committed:
Merge branch 'main' into ASSETS-1126-integrate-aggregated-balance-selector

bergarces commented on ui/pages/multichain-accounts/account-list/account-list.tsx:
```diff
@@ -114,6 +116,13 @@ export const AccountList = () => {
             data-testid="multichain-account-list-search"
           />
         </Box>
+        <MultichainAccountList
```
Seems like a duplicate. Must have happened during a merge. I'll clean it, check it works and push.

bergarces committed:
remove duplicates

salimtb committed:
fix: handle undefined balance case

salimtb committed:
Merge branch 'main' into ASSETS-1126-integrate-aggregated-balance-selector

salimtb committed:
fix: fallback to empty string if balance is not available
